setup.py

```diff
@@ -446,6 +446,9 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
     'pytest',
     'pytest-cov',
     'pytest-instafail',
+    'pytest-rerunfailures',
```

Comment:
Won't be useful to install it I think unless we add the rerun param to the test. Example:

```
@pytest.mark.flaky(reruns=5, reruns_delay=2)
```

Unless there is a default I don't know

Reply:
Yes. I plan to add it in a follow-up PR. I did not not want to drag-along all the changes as I was asked to split the PR into smaller and I would be able to only add --rerun option at the very end where I added quarantine option in #8393 . I figured also based on the comment from @turbaszek https://github.com/apache/airflow/pull/8393#discussion_r411108454 that in this case it's better to split to a "adding new requirements" and "using the new requirements".

Do you think @kaxil it shoudl be merged instead? I am ok to do either. But I think it's a bit cleaner this way.

Reply:
Happy to hear if others have different opinions here on how we should deal with "setup.py" updates :).

Reply:
For now I think that (since we also want to update the constraint files) it will be better to split such commits into " add requirements to setup.py" and "use requirements" PR.

Reply:
But I am happy with both approaches.

Reply:
It is difficult to justify why we are adding this new requirement if it is not being used. 

For example, if we won't add a new operator and we split it into 2: (1) Adds required libraries (2) Adds the actual operator, it wouldn't be as readable I think

A single PR that says "This is the change we need" (dependency is just something we need for that change). 

Happy to hear what others think. 

I think that particular comment from @turbaszek was because botocore change was not related to that PR so I understand his concern and I feel the same but here both would be related.

I am happy either way but happy to hear thoughts

Reply:
And BTW in the quarantine case I am using it as `--reruns` command line parameter for all tests marked as quarantine, so it's really nice :)

Reply:
`reruns` should definitely help :)

Reply:
One thing that is important here. Right now when we change setup.py and you generate requirements.txt files also "unrelated" libraries will get bump (i am using eager update strategy). This is so that the requirements get bumped to reflect what a "new user" will get when the user runs "pip install airflow"  (note that those are not real "requirements" - those are constraint files and their meaning is "those are the set of requirements that we know will work". 

So always when you update setup.py (and only then) you will get as result a bunch of changes to unrelated library versions - we use so many libraries that pretty much every day some library will get released). Now this is good that we have it (this way we can keep being updated with "new user" experience of airflow installation (and fix potential problems early). But then really the "requirements" updates are now connected with "setup.py" changes. and not with the actual functionality using it. 

So we have two options now:

1) Treat "setup.py" updates as an opportunity to bump constraint files and have it a separate commit/PR

2) Merge it all together - including code changes related to the new requirement with unrelated constraint upgrades 

I think 1) is a bit cleaner...

We can also think about 

3) split unrelated constraint upgrades to separate commit - but I think this is much worse than 1) or 2) 

@turbaszek @dimberman @ashb @mik-laj Others?

Reply:
I like (1) too

Reply:
> add requirements to setup.py" and "use requirements" PR.

From a git history point-of-view I'd rather that adding the requirement to setup.py and using it was in the same commit, otherwise cherry-picking will be almost impossible to get right (you have to pick the version update, and the change it needs).

Not to mention that makes it rather more complex for contributors to add new modules -- they'd have to make a PR to add the version, wait for tests to pass, get it merged, then make the change they actually wanted. For that reason alone I think this is the wrong approach -- we should add to setup.py and use it in the same PR. And we'll just have to live with the requirements files changing.

Sidenote: my ideal solution isn't possible without rewriting something like pipenv/poetry:

Thinking about how `yarn` handles this: adding a new requirement will only add this single new dep and it's dependencies, to the lock file, and it's only running `yarn update` or similar that would update all the other versions.

Reply:
I subscribe to the idea that "a commit/PR should do one thing" - so (3), if it worked and was maintainable, would be my ideal.

Reply:
> From a git history point-of-view I'd rather that adding the requirement to setup.py and using it was in the same commit, otherwise cherry-picking will be almost impossible to get right (you have to pick the version update, and the change it needs).

True. That convinces me that we should go with (2). Since the requirement files are just a hint, not the real requirements I don't think it's a bit deal. The requirement files will be re-generated anyway regularly - with every setup.py, so this is really a by-product. I would not worry too much about having unrelated requirement changes when setup.py changes.

Yes. Agree (3) is ideal, but it's not easy.

When you first look at it - (3) is not as "bad"  for developers as it sounds.

It does require a bit of discipline when adding new requirements to setup,py. But It's perfectly workable even now and requires just few specific steps to follow:

1) add/modify requirement in setup.py 
2) Run ./breeze generate-requirements --python [3.6,3.7]
3) git add -p .  adding all other changes but the newly added/changed requirements
4) commit 
5) add/commit the rest of your change

However you would have to do it just before merge, because then if it takes few days your requirements will change again :( or you would have to do fixups to the other commit and rebase two of them. Sounds messy.